shell/platform/fuchsia/flutter/tests/fake_flatland_unittests.cc

```diff
@@ -0,0 +1,129 @@
+  EXPECT_EQ(fake_flatland().debug_name(), "");
+}
+
+TEST_F(FakeFlatlandTest, DebugLabel) {
+  const std::string debug_label = GetCurrentTestName();
+  fuchsia::ui::composition::FlatlandPtr flatland = CreateFlatland();
+
+  // Set the flatland's debug name.  The `SetDebugName` hasn't been processed
+  // yet, so the session's view of the debug name is still empty.
+  flatland->SetDebugName(debug_label);
+  EXPECT_EQ(fake_flatland().debug_name(), "");
+
+  // Pump the loop; the contents of the initial `SetDebugName` should be
+  // processed.
+  loop().RunUntilIdle();
+  EXPECT_EQ(fake_flatland().debug_name(), debug_label);
+}
+
+TEST_F(FakeFlatlandTest, Present) {
+  fuchsia::ui::composition::FlatlandPtr flatland = CreateFlatland();
+
+  // Fire an OnNextFrameBegin event and verify the test receives it.
+  constexpr uint64_t kOnNextFrameAdditionalPresentCredits = 10u;
+  bool on_next_frame_begin_called = false;
+  fuchsia::ui::composition::OnNextFrameBeginValues on_next_frame_begin_values;
+  on_next_frame_begin_values.set_additional_present_credits(
+      kOnNextFrameAdditionalPresentCredits);
+  flatland.events().OnNextFrameBegin =
+      [&on_next_frame_begin_called,
+       kOnNextFrameAdditionalPresentCredits](auto on_next_frame_begin_values) {
+        EXPECT_FALSE(on_next_frame_begin_called);
+        EXPECT_EQ(on_next_frame_begin_values.additional_present_credits(),
+                  kOnNextFrameAdditionalPresentCredits);
+        on_next_frame_begin_called = true;
+      };
+  fake_flatland().FireOnNextFrameBeginEvent(
+      std::move(on_next_frame_begin_values));
+  loop().RunUntilIdle();
+  EXPECT_TRUE(on_next_frame_begin_called);
+
+  // Fire an OnFramePresented event and verify the test receives it.
+  constexpr uint64_t kOnFramePresentedNumPresentsAllowed = 20u;
+  bool on_frame_presented_called = false;
+  flatland.events().OnFramePresented =
+      [&on_frame_presented_called,
+       kOnFramePresentedNumPresentsAllowed](auto frame_presented_info) {
+        EXPECT_FALSE(on_frame_presented_called);
+        EXPECT_EQ(frame_presented_info.num_presents_allowed,
+                  kOnFramePresentedNumPresentsAllowed);
+        on_frame_presented_called = true;
+      };
+  fake_flatland().FireOnFramePresentedEvent(
+      fuchsia::scenic::scheduling::FramePresentedInfo{
+          .actual_presentation_time = 0,
+          .num_presents_allowed = kOnFramePresentedNumPresentsAllowed,
+      });
+  loop().RunUntilIdle();
+  EXPECT_TRUE(on_frame_presented_called);
+
+  // Call Present and verify the fake handled it.
+  constexpr int64_t kPresentRequestedTime = 42;
+  bool present_called = false;
+  fuchsia::ui::composition::PresentArgs present_args;
+  present_args.set_requested_presentation_time(kPresentRequestedTime);
+  fake_flatland().SetPresentHandler(
+      [&present_called, kPresentRequestedTime](auto present_args) {
+        EXPECT_FALSE(present_called);
+        EXPECT_EQ(present_args.requested_presentation_time(),
```

Comment:
Baking EXPECT calls into the callback spreads out the assertion logic in various places. Could we collect the all expected state outside the callback, and have a block of EXPECTs at the bottom? I think it's not a big change, just declare the state to capture, like "std::optional<Foo> foo;", and closure-capture foo in the callback. Here and above.

Reply:
Done